### PR TITLE
Fix llama sin_cached/cos_cached backward compatibility

### PR DESCRIPTION
**Pull request by fxmarty**
The `_sin_cached` & `_cos_cached` are never set in the init (compare to https://github.com/huggingface/transformers/blob/v4.37.2/src/transformers/models/llama/modeling_llama.py#L134-L136), which yields errors in external packages as backward compatibility is broken (e.g. in https://github.com/AutoGPTQ/AutoGPTQ/blob/6b55300dd83326504ee6e02b730fa4451adfa479/auto_gptq/modeling/_utils.py#L95-L96)

IMO this should be in a patch release.

### CODE REVIEW

File: src/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -100,6 +100,21 @@ def __init__(self, dim, max_position_embeddings=2048, base=10000, device=None):
         inv_freq = 1.0 / (self.base ** (torch.arange(0, self.dim, 2, dtype=torch.int64).float().to(device) / self.dim))
         self.register_buffer("inv_freq", inv_freq, persistent=False)
 
+        # TODO: Remove in 4.40.
```
Comment (amyeroberts) [inline src/transformers/models/llama/modeling_llama.py:103]: why 4.40 here? This kind of version dependant removal would be for deprecation of a feature, but AFAICT in the PR comment we don't have an implemented fix which replaces this

Comment (amyeroberts) [reply src/transformers/models/llama/modeling_llama.py:103]: huh - OK. Won't this means things still break though?

**Comment by gante** [comment]
@fxmarty The extent of the fix may depend on the following question: are the libraries downstream broken because a) of the lack of the tensors, or because b) the lack of the tensors AND their values?

The PR as it stands would fix a), but it probably wouldn't fix b). 

_________________________________________
Full story of how this came to be:
1. Static cache removed these tensors
2. [This PR](https://github.com/huggingface/transformers/pull/29109) added them, but a) NOT at init time (which this PR would fix as of the latest commit) and b) the contents and shape of the tensors are different 

Related PR: https://github.com/huggingface/transformers/pull/29198 (which tries to fix https://github.com/huggingface/transformers/issues/29173)

**Comment by gante** [comment]
Note to ourselves: non-permanent buffers can't be treated as common variables for deprecation purposes 😬

File: src/transformers/models/llama/modeling_llama.py (hunk 1 above)
Comment (gante) [reply src/transformers/models/llama/modeling_llama.py:103]: I don't think we can remove them, no 💔

**Comment by ArthurZucker** [comment]
#29198 will add them at init time. 
We keep the values but we deprecate fast because the values are not longer updated.

**Review by ArthurZucker — COMMENTED** [review]
Let's not duplicate the work

**Comment by fxmarty** [comment]
Was not aware https://github.com/huggingface/transformers/pull/29198 was a fix for that. Note that with `self.register_buffer("_cos_cached", ...)` I had some torch.fx tests failing.

**Comment by ArthurZucker** [comment]
Feel free to comment over there

**Pull request closed without merging**